File: foodie-spa/src/services/restaurantService.js
import api from './api';

const BASE_URL = 'http://localhost:8082/api/v1';

class RestaurantService {
  async getAllRestaurants() {
    try {
      const response = await api.get(`${BASE_URL}/restaurants`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch restaurants');
    }
  }

  async getPopularRestaurants(limit = 6) {
    try {
      const response = await api.get(`${BASE_URL}/restaurants?popular=true&limit=${limit}`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch popular restaurants');
    }
  }

  async getOwnerRestaurants() {
    try {
      const response = await api.get(`${BASE_URL}/owner/restaurants`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch owner restaurants');
    }
  }

  async getRestaurantById(id) {
    try {
      const response = await api.get(`${BASE_URL}/restaurants/${id}`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch restaurant');
    }
  }

  async searchRestaurants(query) {
    try {
      const response = await api.get(`${BASE_URL}/restaurants/search?query=${encodeURIComponent(query)}`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Search failed');
    }
  }

  async getRestaurantMenu(restaurantId) {
    try {
      const response = await api.get(`${BASE_URL}/restaurants/${restaurantId}/menu`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch menu');
    }
  }

  async getCuisines() {
    try {
      const response = await api.get(`${BASE_URL}/cuisines`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch cuisines');
    }
  }

  async getPopularCuisines() {
    try {
      const response = await api.get(`${BASE_URL}/cuisines/popular`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch popular cuisines');
    }
  }

  async createRestaurant(restaurantData) {
    try {
      const response = await api.post(`${BASE_URL}/restaurants`, restaurantData);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to create restaurant');
    }
  }

  async updateRestaurant(id, restaurantData) {
    try {
      const response = await api.put(`${BASE_URL}/restaurants/${id}`, restaurantData);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to update restaurant');
    }
  }

  async getOwnerRestaurants() {
    try {
      const response = await api.get(`${BASE_URL}/restaurants/owner`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch owner restaurants');
    }
  }

  async addMenuItem(restaurantId, menuItem) {
    try {
      const response = await api.post(`${BASE_URL}/restaurants/${restaurantId}/menu`, menuItem);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to add menu item');
    }
  }

  async updateMenuItem(restaurantId, itemId, menuItem) {
    try {
      const response = await api.put(`${BASE_URL}/restaurants/${restaurantId}/menu/${itemId}`, menuItem);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to update menu item');
    }
  }

  async deleteMenuItem(restaurantId, itemId) {
    try {
      const response = await api.delete(`${BASE_URL}/restaurants/${restaurantId}/menu/${itemId}`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to delete menu item');
    }
  }
}

export default new RestaurantService();

File: foodie-backend/restaurant-service/src/main/java/com/foodieapp/restaurant/service/RestaurantService.java
package com.foodieapp.restaurant.service;

import com.foodieapp.restaurant.dto.response.RestaurantDTO;
import com.foodieapp.restaurant.dto.response.RestaurantDetailResponse;
import com.foodieapp.restaurant.model.MenuItem;
import com.foodieapp.restaurant.model.Restaurant;

import java.util.List;

public interface RestaurantService {
    // Restaurant operations
    RestaurantDetailResponse createRestaurant(Restaurant restaurant);
    Restaurant getRestaurantById(String id);
    RestaurantDetailResponse getRestaurantDetailById(String id);
    List<RestaurantDetailResponse> getAllRestaurants();
    List<RestaurantDetailResponse> getRestaurantsByOwner(String ownerId);
    List<RestaurantDetailResponse> getRestaurantsByCuisine(String cuisine);
    RestaurantDetailResponse updateRestaurant(String id, Restaurant restaurant);
    void deleteRestaurant(String id);

    // Menu operations
    MenuItem addMenuItem(String restaurantId, MenuItem menuItem);
    MenuItem getMenuItem(String restaurantId, String itemId);
    MenuItem updateMenuItem(String restaurantId, String itemId, MenuItem menuItem);
    void deleteMenuItem(String restaurantId, String itemId);
    List<MenuItem> getMenuItems(String restaurantId);

    // Special queries
    List<RestaurantDetailResponse> getPopularRestaurants(int limit);
    List<RestaurantDetailResponse> getNearbyRestaurants(double latitude, double longitude, double radiusInKm, int limit);
    List<RestaurantDetailResponse> getNewRestaurants(int limit);
    List<RestaurantDetailResponse> getTopRatedRestaurants(int limit);
    List<RestaurantDTO> getFeaturedRestaurants();

    // Note: getPopularCuisines() is not part of this interface
    // It's implemented in RestaurantServiceImpl but not part of the contract
}
